File: app/src/main/java/com/ramez/shopp/RootApplication.java
package com.ramez.shopp;

import android.app.Application;
import android.content.Context;
import android.content.res.Configuration;
import android.util.Log;

import androidx.appcompat.app.AppCompatDelegate;
import androidx.multidex.MultiDex;

import com.androidnetworking.AndroidNetworking;
import com.facebook.drawee.backends.pipeline.Fresco;
import com.google.android.libraries.places.api.Places;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.onesignal.OneSignal;
import com.ramez.shopp.ApiHandler.DataFeacher;
import com.ramez.shopp.Classes.Constants;
import com.ramez.shopp.Classes.ExampleNotificationOpenedHandler;
import com.ramez.shopp.Classes.UtilityApp;
import com.ramez.shopp.Models.MemberModel;
import com.ramez.shopp.Models.ResultAPIModel;
import com.ramez.shopp.Utils.LocaleUtils;
import com.ramez.shopp.Utils.SharedPManger;

import java.util.Locale;


public class RootApplication extends Application {

    private static RootApplication rootApplication;
    private static final String ONESIGNAL_APP_ID = "06c038db-2891-4e93-b03f-ac3a308efc8e";
    public   static  FirebaseAnalytics firebaseAnalytics;

    SharedPManger sharedPManger;


    static {
        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);
    }

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        MultiDex.install(this);
    }

    public static synchronized RootApplication getInstance() {
        return rootApplication;
    }

    public SharedPManger getSharedPManger() {
        return sharedPManger;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        rootApplication = this;
        sharedPManger = new SharedPManger(this);
        firebaseAnalytics = FirebaseAnalytics.getInstance(this);
        Fresco.initialize(this);

        Places.initialize(getApplicationContext(), this.getString(R.string.mapKey), Locale.US);

        OneSignal.setLogLevel(OneSignal.LOG_LEVEL.VERBOSE, OneSignal.LOG_LEVEL.NONE);

        OneSignal.init(this,"",ONESIGNAL_APP_ID);

        OneSignal.startInit(this)
                .setNotificationOpenedHandler(new ExampleNotificationOpenedHandler(this))
                .inFocusDisplaying(OneSignal.OSInFocusDisplayOption.Notification)
                .init();
        String UUID = OneSignal.getPermissionSubscriptionState().getSubscriptionStatus().getUserId();
        if (UUID != null) {
            UtilityApp.setFCMToken(UUID);

        }


        if (UtilityApp.getUserData() != null && UtilityApp.getUserData().getId() != null) {
            String updateToken = OneSignal.getPermissionSubscriptionState().getSubscriptionStatus().getUserId();
            MemberModel memberModel = UtilityApp.getUserData();
            memberModel.setDeviceToken(updateToken);
            UtilityApp.setUserData(memberModel);
            UpdateToken(memberModel);
        }



        String appLanguage = UtilityApp.getLanguage();
        if (appLanguage == null) {
            appLanguage = Constants.English;

        }
        UtilityApp.setLanguage(appLanguage);
        LocaleUtils.setLocale(new Locale(appLanguage));
        LocaleUtils.updateConfig(rootApplication, rootApplication.getResources().getConfiguration());

        AndroidNetworking.initialize(getApplicationContext());


    }


    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        LocaleUtils.updateConfig(rootApplication, newConfig);
    }

    private void UpdateToken(MemberModel memberModel) {

        new DataFeacher(false, (obj, func, IsSuccess) -> {
            ResultAPIModel<String> result = (ResultAPIModel) obj;
            if (IsSuccess) {
                Log.i("TAG", "Log  UpdateToken Success ");


            }


        }).UpdateTokenHandle(memberModel);

    }

}

File: app/src/main/java/com/ramez/shopp/RootApplication.kt
package com.ramez.shopp

import android.app.Application
import android.content.Context
import android.content.res.Configuration
import android.util.Log
import androidx.appcompat.app.AppCompatDelegate
import androidx.multidex.MultiDex
import com.androidnetworking.AndroidNetworking
import com.google.android.libraries.places.api.Places
import com.google.firebase.FirebaseApp
import com.google.firebase.analytics.FirebaseAnalytics
import com.onesignal.OneSignal
import com.ramez.shopp.ApiHandler.DataFeacher
import com.ramez.shopp.ApiHandler.DataFetcherCallBack
import com.ramez.shopp.classes.Constants
import com.ramez.shopp.classes.ExampleNotificationOpenedHandler
import com.ramez.shopp.classes.UtilityApp
import com.ramez.shopp.Models.MemberModel
import com.ramez.shopp.Utils.LocaleUtils
import com.ramez.shopp.Utils.SharedPManger
import java.util.*


class RootApplication : Application() {
    var sharedPManger: SharedPManger? = null

    companion object {
        @get:Synchronized
        var instance: RootApplication? = null
            private set
        private const val ONESIGNAL_APP_ID = "06c038db-2891-4e93-b03f-ac3a308efc8e"
        var firebaseAnalytics: FirebaseAnalytics? = null

        init {
            AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
        }
    }

    override fun attachBaseContext(base: Context) {
        super.attachBaseContext(base)
        MultiDex.install(this)
    }

    override fun onCreate() {
        super.onCreate()
        instance = this
        sharedPManger = SharedPManger(this)
        FirebaseApp.initializeApp(this);
        firebaseAnalytics = FirebaseAnalytics.getInstance(this)
        Places.initialize(applicationContext, this.getString(R.string.mapKey), Locale.US)

        OneSignal.setLogLevel(OneSignal.LOG_LEVEL.VERBOSE, OneSignal.LOG_LEVEL.NONE)

        OneSignal.setLogLevel(OneSignal.LOG_LEVEL.VERBOSE, OneSignal.LOG_LEVEL.NONE);

        // OneSignal Initialization
        OneSignal.initWithContext(this);
        OneSignal.setAppId(ONESIGNAL_APP_ID);
        OneSignal.setNotificationOpenedHandler(ExampleNotificationOpenedHandler(this))

        val UUID = OneSignal.getDeviceState()?.userId
        val UUID1 = OneSignal.getDeviceState()?.pushToken
    
        if (UUID != null) {
            UtilityApp.setFCMToken(UUID)
        }
        Log.d("debug", "Log token one signal   :$UUID")
        Log.d("debug", "Log token one signal2   :$UUID1")

        if (UtilityApp.getUserData() != null && UtilityApp.getUserData().id != null) {
            val updateToken = OneSignal.getDeviceState()?.userId
            val memberModel = UtilityApp.getUserData()
            memberModel.deviceToken = updateToken
            UtilityApp.setUserData(memberModel)
            updateToken(memberModel)
        }
        var appLanguage = UtilityApp.getLanguage()
        if (appLanguage == null) {
            appLanguage = Constants.English
        }
        UtilityApp.setLanguage(appLanguage)
        LocaleUtils.setLocale(Locale(appLanguage))
        LocaleUtils.updateConfig(instance, instance!!.resources.configuration)
        AndroidNetworking.initialize(applicationContext)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        LocaleUtils.updateConfig(instance, newConfig)
    }

    private fun updateToken(memberModel: MemberModel) {
        DataFeacher(false, object : DataFetcherCallBack {
            override fun Result(obj: Any?, func: String?, IsSuccess: Boolean) {
                if (IsSuccess) {
                    Log.i("TAG", "Log  UpdateToken Success ")
                }
            }

        }).UpdateTokenHandle(memberModel)
    }
}

